portlayer/vsphere/storage/store.go

```diff
@@ -226,24 +242,132 @@ func (v *ImageStore) WriteImage(ctx context.Context, parent *portlayer.Image, ID
 		Parent:   parent.SelfLink,
 		Store:    parent.Store,
+		Metadata: meta,
 	}
 
 	return newImage, nil
 }
 
 func (v *ImageStore) GetImage(ctx context.Context, store *url.URL, ID string) (*portlayer.Image, error) {
-	return nil, fmt.Errorf("not yet implemented")
+
+	storeName, err := util.StoreName(store)
+	if err != nil {
+		return nil, err
+	}
+
+	imageURL, err := util.ImageURL(storeName, ID)
+	if err != nil {
+		return nil, err
+	}
+
+	p := v.imageDirPath(storeName, ID)
+	info, err := v.s.Datastore.Stat(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	_, ok := info.(*types.FolderFileInfo)
+	if !ok {
+		return nil, fmt.Errorf("Stat error:  image doesn't exist (%s)", p)
+	}
+
+	meta, err := v.getMeta(ctx, storeName, ID)
+	if err != nil {
+		return nil, err
+	}
+
+	var s url.URL = *store
+
+	newImage := &portlayer.Image{
+		ID:       ID,
+		SelfLink: imageURL,
+		// We're relying on the cache for this since we don't currently have a
+		// way to get the disk's spec.  See VIC #482 for details.  Parent:
+		// parent.SelfLink,
+		Store:    &s,
+		Metadata: meta,
+	}
+
+	return newImage, nil
 }
 
 func (v *ImageStore) ListImages(ctx context.Context, store *url.URL, IDs []string) ([]*portlayer.Image, error) {
 	return nil, fmt.Errorf("not yet implemented")
 }
 
+// Write the opaque metadata blobs (by name) for an image.  We create a
+// directory under the image's parent directory.  Each blob in the metadata map
+// is written to a file with the corresponding name.  Likewise, when we read it
+// back (on restart) we populate the map accordingly.
+func (v *ImageStore) writeMeta(ctx context.Context, storeName string, ID string,
+	meta map[string][]byte) error {
+	// XXX this should be done via disklib so this meta follows the disk in
+	// case of motion.
+
+	metaDataDir := v.imageMetadataDirPath(storeName, ID)
+
+	if meta != nil && len(meta) != 0 {
+		for name, value := range meta {
+			r := bytes.NewReader(value)
+			pth := path.Join(metaDataDir, name)
+			log.Infof("Writing metadata %s", pth)
+			if err := v.s.Datastore.Upload(ctx, r, pth, &soap.DefaultUpload); err != nil {
+				return err
+			}
+		}
+	} else {
+		if err := v.fm.MakeDirectory(ctx, v.datastorePath(metaDataDir), nil, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (v *ImageStore) getMeta(ctx context.Context, storeName string, ID string) (map[string][]byte, error) {
+	metaDataDir := v.datastorePath(v.imageMetadataDirPath(storeName, ID))
+	res, err := lsDir(ctx, v.s.Datastore, metaDataDir)
+	if err != nil {
+		return nil, err
+	}
+
+	meta := make(map[string][]byte)
+	for _, f := range res.File {
+		finfo, ok := f.(*types.FileInfo)
+		if !ok {
+			continue
+		}
+
+		p := path.Join(v.imageMetadataDirPath(storeName, ID), finfo.Path)
+		log.Infof("Getting meta for image (%s) %s", ID, p)
+		rc, _, err := v.s.Datastore.Download(ctx, p, &soap.DefaultDownload)
+		if err != nil {
+			return nil, err
+		}
+		defer rc.Close()
+
+		buf, err := ioutil.ReadAll(rc)
```

Comment:
I agree with the article, but in this case we need the blobs in memory.  We don't want to keep the reader open because the http (govmomi) client remains open to the server and we don't want to buffer the data to disk.  We keep the blobs in memory to avoid the unnecessary datastore operations.